ast-checks/src/missing-patch-comment.rs

```diff
@@ -0,0 +1,104 @@
+    Ok(())
+}
+
+fn analyze_single_file(
+    files: &Files<String>,
+    file_id: FileId,
+) -> Result<Vec<MissingCommentMessage>, Box<dyn Error>> {
+    let ast = rnix::parse(files.source(file_id)).as_result()?;
+    let root = ast.root().inner().ok_or("No elements")?;
+    let mut report: Vec<MissingCommentMessage> = vec![];
+
+    // Find a key in an attrset called 'patches'
+    for kv in walk_keyvalues_filter_key(&root, "patches") {
+        // Look for a comment directly above the `patches = ...` line
+        if let Some(_comment) = find_comment_above(kv.node()) {
+            // Okay!, we found a comment directly above this line,
+            // so I'm interpreting that as the comment which applies to
+            // all patches. It would be _better_ to put a single comment
+            // per patch, but we'll let this slide.
+            continue;
+        }
+
+        // Try to convert the value to a list
+        match kv.value().and_then(List::cast) {
+            Some(value) => {
+                // For each element in the list of patches, look for
+                // a comment directly above the element or, if we don't see
+                // one of those, look for a comment within AST of the element.
+                for item in value.items() {
+                    let has_comment_above = find_comment_above(&item).is_some();
+                    let has_comment_within = find_comment_within(&item).is_some();
+
+                    if !(has_comment_above || has_comment_within) {
+                        let start = item.text_range().start().to_usize() as u32;
+                        let loc = files.location(file_id, start)?;
+
+                        report.push(MissingCommentMessage {
+                            cond: true,
+                            msg: "Please add a comment on the line above explaining the purpose of this patch.",
+                            name: "missing-patch-comment",
+                            locations: vec![SourceLocation {
+                                file: files
+                                    .name(file_id)
+                                    .to_str()
+                                    .ok_or("encoding error")?
+                                    .to_string(),
+                                // 1-based to 0-based indexing
+                                column: loc.column.to_usize() + 1,
+                                line: loc.line.to_usize() + 1,
+                            }],
+                        });
+                    }
+                }
+            }
+            None => {
+                let offset = kv.node().text_range().start().to_usize() as u32;
+                let loc = files.location(file_id, offset)?;
+                eprintln!(
+                    "Unexpected format at line {} column {}. patches is not a list",
+                    loc.line, loc.column
+                );
+                process::exit(1);
+            }
+        }
+    }
```

Comment:
I think you could remove one level of nesting here by writing it like this:
```
let value = kv.value.and_then(List::cast).unwrap_or_else(|| {
  let offset ...
  process::exit(1);
});

for item in value.items() { ... }
```

I think with this you could also convert the for loop into an iterator and use `.collect()` to avoid having that mutable vec and pushing in a loop.